### DISCUSSION TITLE
📚 The Typst Library: a collection of documents written in Typst

### DISCUSSION THREAD

**Discussion by rwmpelstilzchen** [typst/typst#4862] [General]
# The idea

I want to make a webpage that presents an inclusive collection of a wide range of documents written in Typst. I can see two main benefits to such a resource:

- A showcase of what Typst is capable of in the wild. This can be of use for current users, potentially future users (who will be hopefully convinced by what they see) and publishers (who will hopefully see there is enough use and demand for Typst support). It isn’t without reason [this question on tex.stackexchange.com](https://tex.stackexchange.com/questions/1319/showcase-of-beautiful-typography-done-in-tex-friends) is one of the highest scoring there.
- As many documents come with source files, this way if you see something you like you can see how it was produced.

I have experience with collecting massive lists of references (see [Wordles of the World](https://rwmpelstilzchen.gitlab.io/wordles/)). My plan is to use [Isotope](https://isotope.metafizzy.co/) (or a similar library) for displaying (using the first page of each document as a thumbnail) and filtering the list. Filtering can be done with hierarchical tags (inclusive within each category, exclusive across categories, [like so](https://codepen.io/desandro/pen/MebyMR)), with categories that include: document language, document type (book, paper, poster, presentation, etc.), disciplines, packages used, etc. Data will be be stored in YAML files, and the page will be statically-generated (one version with fancy JavaScript and an accessible one with no filtering functionality and auto-aligning, animated thumbnails). The code and metadata collection will be FLOSS, publicly available and collaborative through Git.

# The questions

1. I something like this already exists? I couldn’t find anything, but maybe I didn’t look well enough…

2. I want it to have a clear, simple, self-describing name. The best one I could come with is *The Typst Library*, but I’m not sure how well does it conforms to the [branding guidelines](https://typst.app/legal/brand/). Hi @laurmaedje and @reknih, what’s your input on this? 🙂

**Comment by sitandr** [comment]
The only thing similar to this I know my [Typst Examples Book](https://sitandr.github.io/typst-examples-book/book/), although it's main focus are small examples showing different Typst features. There is still has big work to be done, but I don't have enough time and motivation. 

The first place to look for them is probably showcase channel on Discord, though there is everything shown, not just documents.

Futhermore, there were a document collection used for benchmarks somewhere on Discord too.

Collecting the documents may be pretty hard. Your site has to become popular and easy to submit documents to (my book didn't).

Your documents should be easy to explore (one of main issues of my book and, in my opinion, a bit for official template repository too). Maybe there has to be a system of ratings for your users to be able to find really cool documents in tons of primitive ones.

Overall, I think this is cool and useful idea, but would require a lot of effort to make it viable.

I don't think the Team will mind someone trying to do this, they have a lot of other important stuff to care about. Just make sure to write it is not official.

Anyway, I wish you good luck with that project! 

Feel free to ask me anything, I would like to help a bit.

**Reply by rwmpelstilzchen** [reply]
Thanks for your kind words and encouragement, for the Typst Examples Book (which was very useful for me on multiple occasions 🌼) and for the leads on what to look for and where. I will make an account on Discord for this purpose (I avoid walled gardens as much as I can; I wish Typst had an XMPP or \[matrix\] room).

Another source is platforms such as [Codeberg](https://codeberg.org/explore/repos?only_show_relevant=0&sort=recentupdate&q=&language=Typst) and [GitHub](https://github.com/search?q=path%3A%2F.typ%24%2F&type=code). The problem is there are tons of documents to sift through: junk documents (‘hello world’ stuff and tests), packages, and (in the case of GitHub) multiple `.typ` files for the same document (=a single entry in the Library).

**Reply by rwmpelstilzchen** [reply]
> Maybe there has to be a system of ratings for your users to be able to find really cool documents in tons of primitive ones.

I think GitHub stars make a good metric for that. Theoretically I *could* make a rating system specific to the prospective webpage, but:

- GitHub stars already exist.
- Tons of people are on GitHub, marking repos with stars.
- No need for me to write a rating system. Lazy is good ^___^
- No need to make anti-rigging efforts.
- Without a dynamic rating system, the webpage can be statically generated.

**Reply by rwmpelstilzchen** [reply]
Unfortunately, there aren’t many relevant repositories on Codeberg, sourcehut, GitLab, etc., but when I say ‘GitHub stars’ I mean them as well. At first inputting the number of stars to the database will be done manually, but in the future I (or better, someone else…) will make it automatic.

**Reply by reknih** [reply]
I'd caution against using GitHub / forge stars because that will skew the results:

- Towards projects on GitHub, because it has the largest signed in user base who can casually star something they come across
- Against less technical projects that are not hosted on a forge but on a personal website, social media, ResearchGate, ...

**Comment by laurmaedje** [comment]
To add to @sitandr's great answer: I think it definitely sounds like a cool project. We don't have a problem with something like this at all, as long as its clearly visible that it is not official. 

Regarding the name: While I _do_ like how it sounds, it could perhaps be a bit ambigious on what kind of library it is (documents, packages, standard library) and it sounds a bit official. So perhaps a different name would be better.

**Reply by rwmpelstilzchen** [reply]
Great! 😀
Thanks 🙏

Yeah, *Library* indeed has this problem. *Archives* and *Depository* have the problem of the connotation of being stuffy and dusty. *Treasury* is nice, but pompous… Perhaps a different path is a metaphor for a guide that points to what is out there (instead of a metaphor for a place where you put stuff): *Catalogue*, *Guide* (problematic because it sounds like a guide in the sense of ‘a tutorial’), *Register*, *Atlas*, *Index* (multiple conflicting meanings), *Directory* (same), etc.

**Reply by laurmaedje** [reply]
I think independently of the specific word, all those somewhat suffer from sounding quite official due to their descriptive nature. Replacing "The" with "A" would perhaps fix, though I'm not sure it makes for a nice name.

**Reply by rwmpelstilzchen** [reply]
That indefinite noun phrases make suboptimal names due to their very indefinite nature can be tackled by combining a proper noun and an indefinite noun phrase in apposition. Perhaps *‘Typus: a Typst Catalogue’*, where Latin [*Typus*](https://en.wiktionary.org/wiki/typus#Etymology_1) means — among other, related meanings — ‘*(singular only, collectively)* literature, printed material, books’. This is clearly related to *typeset* etymologically, but it is undoubtedly distinct in form.

**Reply by laurmaedje** [reply]
True!

**Comment by alerque** [comment]
Not exactly the same thing, but along similar lines you might have a look at my [Polytype](https://polytype.dev/hieroglyphs/) project. It's not specifically a library for Typst—in fact the goal is to not focus on any particular typesetting engine but instead be a kind of Rosetta stone between them. I'm trying to collect two general category of examples across as many typesetting engines as people are interested in and contribute sources for:

1. Ways to accomplish the same thing in different engines, e.g. the [dropcaps sample](https://polytype.dev/dropcaps/).
1. The different ways engines handle similar inputs, e.g. the [prime symbols sample](https://polytype.dev/prime-symbol/)

Contributions welcome, even if it is for a single engine. Also web UI stuff to better present the samples is welcome too.

**Comment by rochus-keller** [comment]
Some examples of the same representative content rendered in Typst and some of the TeX heritage contenders would indeed be very useful. [alerque](https://github.com/alerque)'s page goes in that direction, but so far only with fragments, not representative documents. Are there any e.g. papers or theses with a rendering in different typesetting systems? My search so far was not successful.

**Reply by rwmpelstilzchen** [reply]
Not that I’m aware of.

**Reply by alerque** [reply]
I'm certainly open to having full length document samples in Polytype. We might need to adjust some things like how the source viewer works (to handle multiple input files per sample, scroll large files well, etc.) and previews work (currently only the first page of PDFs is previewed, the rest you have to see by downloading the file) but those things can be overcome.

I don't know how useful it would be to have *really long* content like book length, but at least long enough samples to show off representative features of document classes and such would be welcome.

**Reply by rochus-keller** [reply]
Just a PDF of each version (same document typeset in Typst vs. e.g. TeX) would be enough and helpful.
